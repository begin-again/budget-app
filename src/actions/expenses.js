import database from '../firebase/firebase';

export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({ id: ref.key, ...expense }));
    });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

/**
 *
 * @param {String} id
 * @param {Object} updates
 */
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// export const startSetExpenses;
// 1. fetch all expense data
// 2. Parse into and array
// 3. dispatch SET_EXPENSES
export const startSetExpenses = () => {
  return (dispatch) => {
    return database
      .ref('expenses')
      .once('value')
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(childSnap => {
          expenses.push({ id: childSnap.key, ...childSnap.val() });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
