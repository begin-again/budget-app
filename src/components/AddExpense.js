import React from 'react';

const AddExpense = () => {
  const template =
    <div>
        Add Expense component
    </div>
        ;
  return template;
};

export default AddExpense;
