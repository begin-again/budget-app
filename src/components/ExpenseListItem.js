import React from 'react';
import { Link } from 'react-router-dom';

export default ({ description, amount, createdAt, id }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>
        {description}
      </h3>
    </Link>
    <p>{amount} &raquo; {createdAt}</p>
  </div>
);
