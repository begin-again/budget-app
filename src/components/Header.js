import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => {
  const template =
    <header>
      <h1>Expensify</h1>
      <NavLink to='/dashboard' activeClassName='is-active' exact>Dashboard</NavLink>
      <NavLink to='/create' activeClassName='is-active'>Create Expense</NavLink>
    </header>
        ;
  return template;
};

export default Header;
