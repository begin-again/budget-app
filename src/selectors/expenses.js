
import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense, i) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
      let textMatch = true;
      if (text) {
        const rx = new RegExp(text, 'i');
        textMatch = rx.test(expense.description);
      }

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') return b.createdAt - a.createdAt;
      return b.amount - a.amount;
    });
};
