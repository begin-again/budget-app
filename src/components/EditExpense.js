import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

const EditExpense = (props) => {
  const template =
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          console.log('updated', expense);
          props.dispatch(editExpense(props.expense.id, expense));
          props.history.push('/');
        }}
      />
      <button onClick={(e) => {
        props.dispatch(removeExpense({ id: props.expense.id }));
        props.history.push('/');
      }}
      >Remove
      </button>
    </div>
        ;
  return template;
};

const mapStateToProps = (state, props) => (
  {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  }
);

export default connect(mapStateToProps)(EditExpense);
