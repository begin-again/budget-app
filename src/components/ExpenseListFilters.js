import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount } from '../actions/filters';

const ExpenseListFilters = (props) => (
  <div>
    <input
      type='text' placeholder='filter by'
      value={props.filters.text}
      onChange={(event) => {
        props.dispatch(setTextFilter(event.target.value));
      }}
    />
    <select
      value={props.filters.sortBy}
      onChange={(e) => {
        if (e.target.value === 'date') {
          props.dispatch(sortByDate());
        } else {
          props.dispatch(sortByAmount());
        }
      }}
    >
      <option value='amount'>Amount</option>
      <option value='date'>Date</option>
    </select>
  </div>
);

const mapStateToProps = (state) => (
  {
    filters: state.filters
  }
);

export default connect(mapStateToProps)(ExpenseListFilters);
