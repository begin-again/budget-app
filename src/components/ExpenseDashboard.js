import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboard = () => {
  const template =
    <div>
      <ExpenseListFilters />
      <ExpenseList />
    </div>
        ;
  return template;
};

export default ExpenseDashboard;
