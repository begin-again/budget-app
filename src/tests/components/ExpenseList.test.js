/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

describe('Expense List Component', () => {
  it('should render expense list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);

    expect(wrapper).toMatchSnapshot();
  });
  it('should render expense list with empty message', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);

    expect(wrapper).toMatchSnapshot();
  });
});
