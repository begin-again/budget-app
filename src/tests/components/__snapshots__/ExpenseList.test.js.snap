// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense List Component should render expense list with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Expenses
      </span>
    </div>
  </div>
</div>
`;

exports[`Expense List Component should render expense list with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <Component
      amount={195}
      createdAt={0}
      description="sweet gum"
      id="1"
      key="1"
      note=""
    />
    <Component
      amount={500}
      createdAt={-345600000}
      description="candy"
      id="2"
      key="2"
      note=""
    />
    <Component
      amount={123400}
      createdAt={345600000}
      description="beer"
      id="3"
      key="3"
      note="stout"
    />
  </div>
</div>
`;
